From: "Paul E. McKenney" <paul.mckenney@linaro.org>
Date: Mon, 13 Aug 2012 11:17:06 -0700
Subject: rcu: Avoid spurious RCU CPU stall warnings
Git-commit: c96ea7cfdd88d0a67c970502bc5313fede34b86b
Patch-mainline: v3.7-rc1
References: bnc#816586

If a given CPU avoids the idle loop but also avoids starting a new
RCU grace period for a full minute, RCU can issue spurious RCU CPU
stall warnings.  This commit fixes this issue by adding a check for
ongoing grace period to avoid these spurious stall warnings.

Reported-by: Becky Bruce <bgillbruce@gmail.com>
Signed-off-by: Paul E. McKenney <paul.mckenney@linaro.org>
Signed-off-by: Paul E. McKenney <paulmck@linux.vnet.ibm.com>
Reviewed-by: Josh Triplett <josh@joshtriplett.org>
Acked-by: Mike Galbraith <mgalbraith@suse.de>

---
 kernel/rcutree.c |    3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

--- a/kernel/rcutree.c
+++ b/kernel/rcutree.c
@@ -627,7 +627,8 @@ static void check_cpu_stall(struct rcu_s
 	j = ACCESS_ONCE(jiffies);
 	js = ACCESS_ONCE(rsp->jiffies_stall);
 	rnp = rdp->mynode;
-	if ((ACCESS_ONCE(rnp->qsmask) & rdp->grpmask) && ULONG_CMP_GE(j, js)) {
+	if (rcu_gp_in_progress(rsp) &&
+	    (ACCESS_ONCE(rnp->qsmask) & rdp->grpmask) && ULONG_CMP_GE(j, js)) {
 
 		/* We haven't checked in, so go dump stack. */
 		print_cpu_stall(rsp);
